File: 1498.number-of-subsequences-that-satisfy-the-given-sum-condition.py
#
# @lc app=leetcode id=1498 lang=python3
#
# [1498] Number of Subsequences That Satisfy the Given Sum Condition
#

# @lc code=start
from typing import List
class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        n = len(nums)
        MOD = 10 ** 9 + 7
        nums.sort()
        cnt = 0
        j, k = 0, n-1
        while j <= k:
            if nums[j] + nums[k] > target:
                k-=1
            else:
                cnt += (pow(2, k - j, MOD))
                j +=1
        return cnt % MOD


# @lc code=end

File: 1498.number-of-subsequences-that-satisfy-the-given-sum-condition.java
/*
 * @lc app=leetcode id=1498 lang=java
 *
 * [1498] Number of Subsequences That Satisfy the Given Sum Condition
 */
import java.util.Arrays;
// @lc code=start
class Solution {
    private static final int MOD = 1_000_000_007;
    private static final int[] pow2 = new int[100_000];
    private static boolean initialized = false;
    private void initialize() {
        if (initialized) return;
        pow2[0] = 1;
        for (int i = 1; i < pow2.length; i++) {
            pow2[i] = (pow2[i - 1] * 2) % MOD;
        }
        initialized = true;
    }

    public int numSubseq(int[] nums, int target) {
        initialize();
        Arrays.sort(nums);
        int left = 0, right = nums.length - 1;
        int count = 0;
        while (left <= right) {
            if (nums[left] + nums[right] <= target) {
                count = (count + pow2[right - left]) % MOD;
                left++;
            } else {
                right--;
            }
        }
        return count;
    }
}
// @lc code=end

